File: Serialization/Manager/Testing/Python/TestNull.py
from vtkmodules.test import Testing as vtkTesting
from vtkmodules.vtkSerializationManager import vtkObjectManager
from vtkmodules.vtkCommonDataModel import vtkPolyData
from vtkmodules.vtkRenderingCore import vtkActor, vtkPolyDataMapper

import vtkmodules.vtkRenderingOpenGL2 # noqa

import json

class TestNull(vtkTesting.vtkTest):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.om = vtkObjectManager()
        self.om.Initialize()

    def test_null_in_state(self):
        mesh_id = self.om.RegisterObject(vtkPolyData())
        self.om.UpdateStatesFromObjects()
        state = json.loads(self.om.GetState(mesh_id))
        self.assertIsNone(state['Points'])

    def test_invoke_with_null_in_args(self):
        actor = vtkActor()
        mapper = vtkPolyDataMapper()
        actor.SetMapper(mapper)
        actor_id = self.om.RegisterObject(actor)
        self.om.UpdateStatesFromObjects()
        self.om.Invoke(actor_id, "SetMapper", "[null]")
        self.assertIsNone(actor.GetMapper())

    def test_invoke_with_null_in_return(self):
        actor = vtkActor()
        actor_id = self.om.RegisterObject(actor)
        self.om.UpdateStatesFromObjects()
        ret = self.om.Invoke(actor_id, "GetMapper", "[]")
        self.assertIsNone(json.loads(ret))

if __name__ == "__main__":
    vtkTesting.main([(TestNull, 'test')])

File: Web/WebAssembly/Testing/JavaScript/testNull.mjs
async function testNullInState() {
  const vtkWASM = await globalThis.createVTKWASM({})
  const remoteSession = new vtkWASM.vtkStandaloneSession();
  let meshId = remoteSession.create("vtkPolyData");
  if (remoteSession.get(meshId).Points !== null) {
    throw new Error("vtkPolyData.Points should be null initially!");
  }
}

async function testInvokeWithNullInArgs() {
  const vtkWASM = await globalThis.createVTKWASM({})
  const remoteSession = new vtkWASM.vtkStandaloneSession();
  let actorId = remoteSession.create("vtkActor");
  remoteSession.invoke(actorId, "SetMapper", [null]);
  if (remoteSession.get(actorId).Mapper !== null) {
    throw new Error("SetMapper with null argument should have set mapper to null!");
  }
}

async function testInvokeWithNullInReturn() {
  const vtkWASM = await globalThis.createVTKWASM({})
  const remoteSession = new vtkWASM.vtkStandaloneSession();
  let actorId = remoteSession.create("vtkActor");
  let ret = remoteSession.invoke(actorId, "GetMapper", []);
  if (ret !== null) {
    throw new Error("GetMapper should have returned null!");
  }
}

const tests = [
  {
    description: "Null in state",
    test: testNullInState,
  },
  {
    description: "Invoke with null in args",
    test: testInvokeWithNullInArgs,
  },
  {
    description: "Invoke with null in return",
    test: testInvokeWithNullInReturn,
  },
];

let exitCode = 0;
for (let test of tests) {
  try {
    await test.test();
    console.log("✓", test.description);
    exitCode |= 0;
  }
  catch (error) {
    console.log("x", test.description);
    console.log(error);
    exitCode |= 1;
  }
}
process.exit(exitCode);
